src/analysis/asymptotics/asymptotics.lean

```diff
@@ -1247,6 +1247,31 @@ end exists_mul_eq
 
 /-! ### Miscellanous lemmas -/
 
+theorem is_O_at_top_of_div_tends_to_finite {α : Type*} [linear_order α] [nonempty α]
+  {f g : α → 𝕜} (hgf : ∀ᶠ x in filter.at_top, g x = 0 → f x = 0) (c : 𝕜)
+  (H : filter.tendsto (f / g) filter.at_top (𝓝 c)) :
+  is_O f g filter.at_top :=
+begin
+  simp only [is_O_iff, filter.eventually_at_top],
+  rw filter.tendsto_iff_eventually at H,
+  have : ∀ᶠ (x : 𝕜) in 𝓝 c, ∥x∥ ≤ ∥c∥ + 1,
+  { rw filter.eventually_iff_exists_mem,
+    refine ⟨metric.ball c 1, metric.ball_mem_nhds c zero_lt_one,
+      λ y hy, norm_le_norm_add_const_of_dist_le (le_of_lt hy)⟩ },
+  obtain ⟨x₀, h⟩ := filter.eventually_at_top.mp hgf,
+  obtain ⟨x₀', h'⟩ := filter.eventually_at_top.mp (H this),
+  refine ⟨∥c∥ + 1, max x₀ x₀', λ x hx, _⟩,
+  rw [ge_iff_le, max_le_iff] at hx,
+  specialize h x hx.1,
+  specialize h' x hx.2,
+  simp only [pi.div_apply, normed_field.norm_div] at h',
+  by_cases hfx : f x = 0,
+  { have : ∥f x∥ = 0 := trans (congr_arg _ hfx) norm_zero,
+    exact this.symm ▸ mul_nonneg (by simpa [hfx] using h') (norm_nonneg (g x)) },
+  { rwa div_le_iff _ at h',
+    exact lt_of_le_of_ne (norm_nonneg (g x)) (λ h', ((mt h) hfx) (norm_eq_zero.mp h'.symm)) }
+end
+
```

Comment:
I think it would be great to split this proof a bit, so we get more general results at the same time. The way I would do it is : 

- prove that if `∥f / g∥` is eventually bounded above along a filter `l` - that is, if `is_bounded_under (≤) l (λ x, ∥(f  x) / (g x)∥)` -, then `is_O f g l`. This should be similar to what you did, but generalized to any filter, not only `at_top` (like @sgouezel said, I think it should be true).
- prove that if a function `f` is eventually bounded from both sides along a filter `l`, then `(λ x, ∥f  x∥)` is eventually bounded above along `l` (this might already exist, but I can't find it)
- then, you can get your result by proving that if `f/g` converges along `l`, then `f/g` is eventually bounded from both sides along `l` (see `filter.tendsto.is_bounded_under_le` and `filter.tendsto.is_bounded_under_ge`), and then using the two results above

If what I say is not clear, or if you're not used to filters and you want some help with what I said, feel free to ask me on Zulip, but I think it would be really great to have those

Reply:
I'm not super familiar with filters, but this did manage to make a lot of the proofs cleaner. I added `is_bounded_under_of_tends_to_nhds` in place of the second suggestion, because `filter.tendsto.is_bounded_under_le` requires some ordering on 𝕜, which the current `is_O_of_div_tends_to_nhds` doesn't assume at all.

Reply:
Oh indeed it is better not to need any ordering, thanks !